File: src/anthropic-admin.ts
/**
 * Anthropic Admin API client — thin, typed, zero-dependency (native fetch).
 *
 * Requires an sk-ant-admin-* key from:
 * https://console.anthropic.com/settings/admin-keys
 */

import type {
  RawApiKeysResponse,
  RawClaudeCodeReport,
  RawCostReport,
  RawMembersResponse,
  RawOrganization,
  RawUsageReport,
  RawWorkspacesResponse,
} from "./types.js";

const BASE_URL = "https://api.anthropic.com";
const API_VERSION = "2023-06-01";

export class AnthropicAdminClient {
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly headers: Record<string, string>;

  constructor(apiKey: string, baseUrl = BASE_URL) {
    if (!apiKey) {
      throw new Error(
        "ANTHROPIC_ADMIN_API_KEY is required. " +
          "Get one at https://console.anthropic.com/settings/admin-keys"
      );
    }
    this.apiKey = apiKey;
    this.baseUrl = baseUrl;
    this.headers = {
      "x-api-key": this.apiKey,
      "anthropic-version": API_VERSION,
      "content-type": "application/json",
    };
  }

  // ── HTTP helpers ─────────────────────────────────────────────

  private async get<T>(
    path: string,
    params?: Record<string, string | number | boolean>
  ): Promise<T> {
    const url = new URL(path, this.baseUrl);
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined && value !== null) {
          url.searchParams.set(key, String(value));
        }
      }
    }

    const response = await fetch(url.toString(), {
      method: "GET",
      headers: this.headers,
    });

    if (!response.ok) {
      const body = await response.text();
      throw new Error(
        `Anthropic Admin API ${response.status}: ${response.statusText} — ${body}`
      );
    }

    return (await response.json()) as T;
  }

  // ── Organization ─────────────────────────────────────────────

  async getOrganization(): Promise<RawOrganization> {
    return this.get<RawOrganization>("/v1/organizations/me");
  }

  // ── Usage Reports ────────────────────────────────────────────

  async getUsageReport(opts: {
    starting_at: string;
    ending_at: string;
    bucket_width?: string;
    group_by?: string;
    model?: string;
    limit?: number;
    page?: string;
  }): Promise<RawUsageReport> {
    return this.get<RawUsageReport>(
      "/v1/organizations/usage_report/messages",
      {
        starting_at: opts.starting_at,
        ending_at: opts.ending_at,
        bucket_width: opts.bucket_width ?? "1d",
        limit: opts.limit ?? 31,
        ...(opts.group_by && { "group_by[]": opts.group_by }),
        ...(opts.model && { model: opts.model }),
        ...(opts.page && { page: opts.page }),
      }
    );
  }

  async getUsageReportByModel(opts: {
    starting_at: string;
    ending_at: string;
    bucket_width?: string;
    limit?: number;
  }): Promise<RawUsageReport> {
    return this.getUsageReport({ ...opts, group_by: "model" });
  }

  // ── Cost Reports ─────────────────────────────────────────────

  async getCostReport(opts: {
    starting_at: string;
    ending_at: string;
    bucket_width?: string;
    group_by?: string;
    limit?: number;
    page?: string;
  }): Promise<RawCostReport> {
    return this.get<RawCostReport>("/v1/organizations/cost_report", {
      starting_at: opts.starting_at,
      ending_at: opts.ending_at,
      bucket_width: opts.bucket_width ?? "1d",
      limit: opts.limit ?? 31,
      ...(opts.group_by && { "group_by[]": opts.group_by }),
      ...(opts.page && { page: opts.page }),
    });
  }

  // ── Claude Code Usage ────────────────────────────────────────

  async getClaudeCodeUsage(opts: {
    starting_at: string;
    limit?: number;
    page?: string;
  }): Promise<RawClaudeCodeReport> {
    return this.get<RawClaudeCodeReport>(
      "/v1/organizations/usage_report/claude_code",
      {
        starting_at: opts.starting_at,
        limit: opts.limit ?? 1000,
        ...(opts.page && { page: opts.page }),
      }
    );
  }

  /**
   * Fetch Claude Code usage for multiple days in parallel.
   * The API requires one call per day.
   */
  async getClaudeCodeUsageRange(
    dates: string[]
  ): Promise<RawClaudeCodeReport[]> {
    const results = await Promise.allSettled(
      dates.map((date) => this.getClaudeCodeUsage({ starting_at: date }))
    );

    const reports: RawClaudeCodeReport[] = [];
    for (const result of results) {
      if (result.status === "fulfilled") {
        reports.push(result.value);
      }
      // Silently skip failed days (matches Python behavior)
    }
    return reports;
  }

  // ── Workspaces ───────────────────────────────────────────────

  async listWorkspaces(opts?: {
    limit?: number;
    include_archived?: boolean;
  }): Promise<RawWorkspacesResponse> {
    return this.get<RawWorkspacesResponse>("/v1/organizations/workspaces", {
      limit: opts?.limit ?? 100,
      include_archived: opts?.include_archived ?? true,
    });
  }

  // ── Members ──────────────────────────────────────────────────

  async listMembers(opts?: {
    limit?: number;
  }): Promise<RawMembersResponse> {
    return this.get<RawMembersResponse>("/v1/organizations/users", {
      limit: opts?.limit ?? 100,
    });
  }

  // ── API Keys ─────────────────────────────────────────────────

  async listApiKeys(opts?: {
    limit?: number;
    status?: string;
  }): Promise<RawApiKeysResponse> {
    return this.get<RawApiKeysResponse>("/v1/organizations/api_keys", {
      limit: opts?.limit ?? 100,
      ...(opts?.status && { status: opts.status }),
    });
  }
}

File: python/src/anthropic_admin.py
"""
Anthropic Admin API client — thin, typed, zero-dependency (urllib.request).

Requires an sk-ant-admin-* key from:
https://console.anthropic.com/settings/admin-keys
"""

from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Any, Optional

BASE_URL = "https://api.anthropic.com"
API_VERSION = "2023-06-01"


class AnthropicAdminClient:
    def __init__(self, api_key: str, base_url: str = BASE_URL) -> None:
        if not api_key:
            raise ValueError(
                "ANTHROPIC_ADMIN_API_KEY is required. "
                "Get one at https://console.anthropic.com/settings/admin-keys"
            )
        self._api_key = api_key
        self._base_url = base_url
        self._headers = {
            "x-api-key": self._api_key,
            "anthropic-version": API_VERSION,
            "content-type": "application/json",
        }

    # -- HTTP helpers ----------------------------------------------------------

    def _get(
        self,
        path: str,
        params: Optional[dict[str, Any]] = None,
    ) -> dict:
        url = f"{self._base_url}{path}"
        if params:
            filtered = {
                k: str(v) for k, v in params.items() if v is not None
            }
            if filtered:
                url = f"{url}?{urllib.parse.urlencode(filtered)}"

        req = urllib.request.Request(url, headers=self._headers, method="GET")

        try:
            with urllib.request.urlopen(req) as resp:
                body = resp.read().decode("utf-8")
                return json.loads(body)
        except urllib.error.HTTPError as e:
            error_body = e.read().decode("utf-8") if e.fp else ""
            raise RuntimeError(
                f"Anthropic Admin API {e.code}: {e.reason} — {error_body}"
            ) from e

    # -- Organization ----------------------------------------------------------

    def get_organization(self) -> dict:
        return self._get("/v1/organizations/me")

    # -- Usage Reports ---------------------------------------------------------

    def get_usage_report(
        self,
        *,
        starting_at: str,
        ending_at: str,
        bucket_width: str = "1d",
        group_by: Optional[str] = None,
        model: Optional[str] = None,
        limit: int = 31,
        page: Optional[str] = None,
    ) -> dict:
        params: dict[str, Any] = {
            "starting_at": starting_at,
            "ending_at": ending_at,
            "bucket_width": bucket_width,
            "limit": limit,
        }
        if group_by:
            params["group_by[]"] = group_by
        if model:
            params["model"] = model
        if page:
            params["page"] = page
        return self._get("/v1/organizations/usage_report/messages", params)

    def get_usage_report_by_model(
        self,
        *,
        starting_at: str,
        ending_at: str,
        bucket_width: str = "1d",
        limit: int = 31,
    ) -> dict:
        return self.get_usage_report(
            starting_at=starting_at,
            ending_at=ending_at,
            bucket_width=bucket_width,
            limit=limit,
            group_by="model",
        )

    # -- Cost Reports ----------------------------------------------------------

    def get_cost_report(
        self,
        *,
        starting_at: str,
        ending_at: str,
        bucket_width: str = "1d",
        group_by: Optional[str] = None,
        limit: int = 31,
        page: Optional[str] = None,
    ) -> dict:
        params: dict[str, Any] = {
            "starting_at": starting_at,
            "ending_at": ending_at,
            "bucket_width": bucket_width,
            "limit": limit,
        }
        if group_by:
            params["group_by[]"] = group_by
        if page:
            params["page"] = page
        return self._get("/v1/organizations/cost_report", params)

    # -- Claude Code Usage -----------------------------------------------------

    def get_claude_code_usage(
        self,
        *,
        starting_at: str,
        limit: int = 1000,
        page: Optional[str] = None,
    ) -> dict:
        params: dict[str, Any] = {
            "starting_at": starting_at,
            "limit": limit,
        }
        if page:
            params["page"] = page
        return self._get("/v1/organizations/usage_report/claude_code", params)

    def get_claude_code_usage_range(self, dates: list[str]) -> list[dict]:
        """Fetch Claude Code usage for multiple days in parallel.

        The API requires one call per day. Failed days are silently skipped
        (mirrors Promise.allSettled semantics in the TypeScript version).
        """
        if not dates:
            return []

        reports: list[dict] = []
        with ThreadPoolExecutor(max_workers=min(len(dates), 10)) as executor:
            futures = {
                executor.submit(self.get_claude_code_usage, starting_at=date): date
                for date in dates
            }
            for future in as_completed(futures):
                try:
                    reports.append(future.result())
                except Exception:
                    pass  # Silently skip failed days
        return reports

    # -- Workspaces ------------------------------------------------------------

    def list_workspaces(
        self,
        *,
        limit: int = 100,
        include_archived: bool = True,
    ) -> dict:
        return self._get("/v1/organizations/workspaces", {
            "limit": limit,
            "include_archived": str(include_archived).lower(),
        })

    # -- Members ---------------------------------------------------------------

    def list_members(self, *, limit: int = 100) -> dict:
        return self._get("/v1/organizations/users", {"limit": limit})

    # -- API Keys --------------------------------------------------------------

    def list_api_keys(
        self,
        *,
        limit: int = 100,
        status: Optional[str] = None,
    ) -> dict:
        params: dict[str, Any] = {"limit": limit}
        if status:
            params["status"] = status
        return self._get("/v1/organizations/api_keys", params)
